pkg/controller/cloud/node_lifecycle_controller.go

```diff
@@ -147,7 +147,38 @@ func (c *CloudNodeLifecycleController) MonitorNodes() {
 			continue
 		}
 
-		// we need to check this first to get taint working in similar in all cloudproviders
+		// At this point the node has NotReady status, we need to check if the node has been removed
+		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node
+		exists, err := ensureNodeExistsByProviderID(context.TODO(), instances, node)
+		if err != nil {
+			klog.Errorf("error checking if node %s exists: %v", node.Name, err)
+			continue
+		}
+
+		if !exists {
+			// Current node does not exist, we should delete it
+
+			klog.V(2).Infof("deleting node since it is no longer present in cloud provider: %s", node.Name)
+
+			ref := &v1.ObjectReference{
+				Kind:      "Node",
+				Name:      node.Name,
+				UID:       types.UID(node.UID),
+				Namespace: "",
+			}
+
+			c.recorder.Eventf(ref, v1.EventTypeNormal,
+				fmt.Sprintf("Deleting node %v because it does not exist in the cloud provider", node.Name),
+				"Node %s event: %s", node.Name, deleteNodeEvent)
+
+			if err := c.kubeClient.CoreV1().Nodes().Delete(context.TODO(), node.Name, metav1.DeleteOptions{}); err != nil {
+				klog.Errorf("unable to delete node %q: %v", node.Name, err)
+			}
+
+			continue
```

Comment:
I'm generally in favor of this but having the `continue` 20 lines down from the initial `if` statement makes it a bit harder to follow. What do you think about removing this `continue` and using an `else` check instead?
```go
if !exists {
    // delete node
} else {
    // check shutdown.
}

Reply:
I think branching is indeed much clearer than using continue. I will remove similar issue in the original code too.